src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,166 @@
+/* @flow */
+import * as React from 'react';
+import { compose } from 'redux';
+
+import log from 'core/logger';
+import translate from 'core/i18n/translate';
+import type { I18nType } from 'core/types/i18n';
+import Permission from 'ui/components/Permission';
+
+/* eslint-disable no-continue */
+type Props = {|
+  i18n: I18nType,
+  permissions: Array<string>,
+|};
+
+type HostPermissionMessageType = 'domain' | 'site';
+type PermissionMessageType = HostPermissionMessageType | 'allUrls';
+
+type GetPermissionStringParams = {|
+  messageType: PermissionMessageType,
+  param?: string | number,
+  tooMany?: boolean,
+|};
+
+type GenerateHostPermissionsParams = {|
+  permissions: Array<string>,
+  // eslint-disable-next-line react/no-unused-prop-types
+  messageType: HostPermissionMessageType,
+|};
+
+export class HostPermissionsBase extends React.Component<Props> {
+  constructor(props: Props) {
+    super(props);
+    this.i18n = props.i18n;
+  }
+
+  getPermissionString({ messageType, param, tooMany = false }: GetPermissionStringParams): string {
+    // These should be kept in sync with Firefox's strings for webextention
+    // host permissions which can be found in
+    // https://hg.mozilla.org/mozilla-central/raw-file/tip/browser/locales/en-US/chrome/browser/browser.properties
+    switch (messageType) {
+      case 'allUrls':
+        return this.i18n.gettext('Access your data for all websites');
+      case 'domain':
+        if (tooMany) {
+          return this.i18n.sprintf(this.i18n.ngettext(
+            'Access your data in %(param)s other domain',
+            'Access your data in %(param)s other domains',
+            param), { param: this.i18n.formatNumber(param) }
+          );
+        }
+        return this.i18n.sprintf(
+          this.i18n.gettext('Access your data for sites in the %(param)s domain'),
+          { param }
+        );
+      case 'site':
+        if (tooMany) {
+          return this.i18n.sprintf(this.i18n.ngettext(
+            'Access your data on %(param)s other site',
+            'Access your data on %(param)s other sites',
+            param), { param: this.i18n.formatNumber(param) }
+          );
+        }
+        return this.i18n.sprintf(
+          this.i18n.gettext('Access your data for %(param)s'),
+          { param }
+        );
+      default:
+        return '';
```

Comment:
This seems like a programmer error and should throw, right? Why would this ever be blank?